### ISSUE TITLE
'wrap_comments' do not support Chinese or Japanese comments.

### ISSUE THREAD

**Issue by garyhai** [rust-lang-nursery/rustfmt#2797]
When set 'wap_comments = true', it works fine for English comments. But when write comments in Chinese or Japanese, the long lines more than comment_width do not be wrapped.

**Comment by wada314** [comment]
I believe this is kinda already recognized problem but I just write this for my note.
TL;DR: This is very difficult problem.

https://github.com/rust-lang-nursery/rustfmt/blob/87edd75ecf26c9084969f431bb5e363693a8a4ca/src/string.rs#L46-L148
https://github.com/rust-lang-nursery/rustfmt/blob/87edd75ecf26c9084969f431bb5e363693a8a4ca/src/utils.rs#L329-L350

Currently rustfmt's line wrapping is not considering much about East-Asian scripts like Japanese, Chinese.
I just took a quick glance but I can point out at least 2 problems:

1. [The styleguide](https://github.com/rust-lang-nursery/fmt-rfcs/blob/master/guide/guide.md) is not saying anything about this but I think neither str::len() nor Chars::count() should be considered as "line length". The string "ゆゆ式" 's str::len() is 9 and Chars::count() is 3 but in normal monospace glyph it occupies 6 width. Unicode has a TR for this: [UAX#11](https://www.unicode.org/reports/tr11/) ([rust impl](https://github.com/unicode-rs/unicode-width)). Though this spec is still leaving us to decide which "context" we should use between CJK and non-CJK contexts. Maybe we can add this setting into rustfmt.toml?

1. Current line-breaking algorithm is not considering into non-space-separating languages (e.g. CJK). The current implementation is partially using Unicode TR [[UAX#29]](http://www.unicode.org/reports/tr29/) ([rust impl](https://github.com/unicode-rs/unicode-segmentation)) but it's not something working for CJK. I also feel like we would rather use [[UAX#14]](https://www.unicode.org/reports/tr14/) to decide the place to insert newline, but unfortunately we don't have official implementation under the official [unicode-rs](https://github.com/unicode-rs) directory today because the TR is not trivial.

So... maybe we should start from implementing [[UAX#14]](https://www.unicode.org/reports/tr14/)? (just kidding)

**Comment by nrc** [comment]
Thanks for the useful info! I'm sorry this still doesn't work. This is covered by #6, so closing this issue, but I do want to fix this.
